relayer/chain/ethereum/listener.go

```diff
@@ -20,47 +20,61 @@ type Listener struct {
 	conn      *Connection
 	contracts []Contract
 	messages  chan<- chain.Message
+	headers   chan<- chain.Header
 	log       *logrus.Entry
 }
 
-func NewListener(conn *Connection, messages chan<- chain.Message, contracts []Contract, log *logrus.Entry) (*Listener, error) {
+func NewListener(conn *Connection, messages chan<- chain.Message, headers chan<- chain.Header, contracts []Contract, log *logrus.Entry) (*Listener, error) {
 	return &Listener{
 		conn:      conn,
 		contracts: contracts,
 		messages:  messages,
+		headers:   headers,
 		log:       log,
 	}, nil
 }
 
 func (li *Listener) Start(cxt context.Context, eg *errgroup.Group) error {
 	eg.Go(func() error {
-		return li.pollEvents(cxt)
+		return li.pollEventsAndHeaders(cxt)
 	})
 
 	return nil
 }
 
-func (li *Listener) pollEvents(ctx context.Context) error {
-	li.log.Info("Polling started")
-
+func (li *Listener) pollEventsAndHeaders(ctx context.Context) error {
 	events := make(chan gethTypes.Log)
-	for _, contract := range li.contracts {
-		query := makeQuery(contract)
+	if li.messages == nil {
+		li.log.Info("Not polling events since channel is nil")
+	} else {
+		li.log.Info("Polling events started")
+
+		for _, contract := range li.contracts {
+			query := makeQuery(contract)
+
+			_, err := li.conn.client.SubscribeFilterLogs(ctx, query, events)
+			if err != nil {
+				li.log.WithFields(logrus.Fields{
+					"address": contract.Address.Hex(),
+				}).Error("Failed to subscribe to application events")
+				continue
+			}
 
-		_, err := li.conn.client.SubscribeFilterLogs(ctx, query, events)
-		if err != nil {
 			li.log.WithFields(logrus.Fields{
-				"address": contract.Address.Hex(),
-			}).Error("Failed to subscribe to application events")
-			continue
+				"contractAddress": contract.Address.Hex(),
+				"contractName":    contract.Name,
+			}).Info("Subscribed to contract events")
 		}
+	}
 
-		li.log.WithFields(logrus.Fields{
-			"contractAddress": contract.Address.Hex(),
-			"contractName":    contract.Name,
-		}).Info("Subscribed to contract events")
+	headers := make(chan *gethTypes.Header)
+	_, err := li.conn.client.SubscribeNewHead(ctx, headers)
```

Comment:
Oh I missed a problem during my previous review. See https://github.com/Snowfork/polkadot-ethereum/pull/163 for more context.

The problem is that background subscriptions created by `li.conn.client.SubscribeNewHead` can fail if the connection to Infura drops for any reason.

Can you add code like the following as a stop-gap measure (or anything else as you see fit)?
https://github.com/Snowfork/polkadot-ethereum/pull/163/files#diff-30d1bf781a4c9c5a1e0134c30fcdf147722200659b16c9cdd4a62b9390e22809R67

```golang
		case err := <-subscription.Err():
			li.log.WithError(err).Error("Subscription terminated")
			return err
```

It will make sure the relayer is terminated after a connection drop (so it can be restarted by Systemd).

Reply:
I'll update this today